tools/afl-fuzzer/main.c

```diff
@@ -0,0 +1,379 @@
+void fuzzer_ipc_crash(struct fuzz *fuzzer, struct mailbox *mailbox,
+		      unsigned int offset)
+{
+	/* TODO: DSP FW has crashed. dump stack, regs, last IPC, log etc */
+	fprintf(stderr, "error: DSP FW crash\n");
+	exit(EXIT_FAILURE);
+}
+
+int fuzzer_send_msg(struct fuzz *fuzzer)
+{
+	struct timespec timeout;
+	struct timeval tp;
+	int ret;
+
+	ipc_dump(&fuzzer->msg);
+
+	/* send msg */
+	ret = fuzzer->platform->send_msg(fuzzer, &fuzzer->msg);
+	if (ret < 0) {
+		fprintf(stderr, "error: message tx failed\n");
+		return ret;
+	}
+
+	/* wait for ipc reply */
+	gettimeofday(&tp, NULL);
+	timeout.tv_sec  = tp.tv_sec;
+	timeout.tv_nsec = tp.tv_usec * 1000;
+	timeout.tv_nsec += 300000000; /* 300ms timeout */
+
+	/* first lock the boot wait mutex */
+	pthread_mutex_lock(&ipc_mutex);
+
+	/* reset condition for next IPC message */
+	ipc_reply_recd = 0;
+	ret = 0;
+
+	/* now wait for mutex unlock and ipc_reply_recd set by IPC reply */
+	while (!ipc_reply_recd && !ret)
+		ret = pthread_cond_timedwait(&ipc_cond, &ipc_mutex, &timeout);
+
+	if (ret == ETIMEDOUT && !ipc_reply_recd) {
+		ret = -EINVAL;
+		fprintf(stderr, "error: IPC timeout\n");
+		ipc_dump_err(&fuzzer->msg);
+		pthread_mutex_unlock(&ipc_mutex);
+		exit(0);
+	}
+
+	pthread_mutex_unlock(&ipc_mutex);
+
+	/*
+	 * sleep for 5 ms before continuing sending the next message.
+	 * This helps with the condition signaling. Without this,
+	 * the condition seems to always satisfy and
+	 * the fuzzer never waits for a response from the DSP.
+	 */
+	usleep(50000);
+
+	return ret;
+}
+
+int main(int argc, char *argv[])
+{
+	struct fuzz fuzzer;
+	int ret;
+	char *ipc_msg_file = NULL;
+	char *platform_name = "byt";
+	char opt;
+	FILE *fp = NULL;
+
+	while ((opt = getopt(argc, argv, "m:")) != -1) {
+		switch (opt) {
+		case 'm':
+			ipc_msg_file = optarg;
+			break;
+		default:
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	if (argc != 2) {
+		fprintf(stderr, "error: Usage ./sof-afl-fuzzer file\n");
+		exit(EXIT_FAILURE);
+	}
+
+	/* init platform */
+	fprintf(stdout, "initialising platform %s\n", platform[0]->name);
+	ret = platform[0]->init(&fuzzer, platform[0]);
+
+	if (ret == ETIMEDOUT) {
+		fprintf(stderr, "error: platform %s failed to initialise\n",
+			platform_name);
+		exit(EXIT_FAILURE);
+	}
+
+	fprintf(stdout, "FW boot complete\n");
+
+	/* allocate max ipc size bytes for the msg and reply */
+	fuzzer.msg.msg_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+	fuzzer.msg.reply_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+
+	/* Open ipc message file */
+	fp = fopen(argv[1], "rb");
+	if (!fp) {
+		fprintf(stderr, "error: opening ipc msg file %s\n",
+			argv[1]);
+		exit(EXIT_FAILURE);
+	}
+
+	/* ipc message file structure:
+	   header
+	   msg_size
+	   msg data
+	*/
+
+	ret = fread(&fuzzer.msg.header, sizeof(uint32_t), 1, fp);
+	
+	if (ret != 1) {
+		fprintf(stderr, "error: reading header of ipc message from %s failed\n",
+			argv[1]);
+		exit(EXIT_FAILURE);
+	}
+
+	fprintf(stdout, "header read: %ud\n", fuzzer.msg.header);
```

Comment:
I think we should dump the output into a separate output file

Reply:
Or, we can have a static debug variable. Only when it's set to 1, these statements are printed to stdout (as we are using these statements to see what the ipc message contains, it'll come under debug right?).